File: apps/mobile/modules/podverse-media-engine/android/src/main/java/expo/modules/podversemediaengine/PodverseVideoSurfaceView.kt
package expo.modules.podversemediaengine

import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

// The RN-mounted host for the ONE shared video surface. Mini player and full player each mount one
// `PodverseVideoSurfaceView` with a `targetId`; the single `SurfaceView` owned by
// `PodverseVideoSurfaceHost` is **reparented** into whichever view is the active target. Because
// these views live inside the RN tree — the full player is a React Navigation native-stack modal
// (react-native-screens fragment / its own window) — the surface renders correctly in both states.
// The process uses one ExoPlayer and one SurfaceView.
class PodverseVideoSurfaceView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext) {
  private var targetId: String? = null

  /** `targetId` prop from JS (`mini` / `full`). Re-registers with the host when it changes. */
  fun setTargetId(value: String) {
    if (value == targetId) return
    targetId?.let { PodverseVideoSurfaceHost.unregisterTargetView(it, this) }
    targetId = value
    if (isAttachedToWindow) {
      PodverseVideoSurfaceHost.registerTargetView(value, this)
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    targetId?.let { PodverseVideoSurfaceHost.registerTargetView(it, this) }
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    targetId?.let { PodverseVideoSurfaceHost.unregisterTargetView(it, this) }
  }

  override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
    super.onLayout(changed, left, top, right, bottom)
    // Observe size changes (rotation, mini↔full) so the reparented surface fills us.
    targetId?.let { PodverseVideoSurfaceHost.layoutSurface(this, it) }
  }
}

File: apps/mobile/modules/podverse-media-engine/ios/PodverseVideoSurfaceView.swift
import ExpoModulesCore
import UIKit

// The RN-mounted host for the ONE shared video surface. Mini player and full player each mount one
// `PodverseVideoSurfaceView` with a `targetId`; the single `AVPlayerLayer`-backed surface owned by
// `PodverseVideoSurfaceHost` is **reparented** into whichever view is the active target. Because
// these views live inside the RN tree — the full player is a React Navigation native-stack modal
// (its own view controller / z-order) — the surface renders correctly in both states. The process
// uses one player and one layer.
public final class PodverseVideoSurfaceView: ExpoView {
  private var targetId: PodverseVideoTargetId?

  public required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    isUserInteractionEnabled = false
    backgroundColor = .clear
  }

  /// `targetId` prop from JS (`mini` / `full`). Re-registers with the host when it changes.
  func setTargetId(_ raw: String) {
    let parsed = PodverseVideoTargetId(rawValue: raw)
    guard parsed != targetId else { return }
    if let previous = targetId {
      PodverseVideoSurfaceHost.shared.unregisterTargetView(previous, view: self)
    }
    targetId = parsed
    if let parsed = parsed, window != nil {
      PodverseVideoSurfaceHost.shared.registerTargetView(parsed, view: self)
    }
  }

  public override func didMoveToWindow() {
    super.didMoveToWindow()
    guard let targetId = targetId else { return }
    if window != nil {
      PodverseVideoSurfaceHost.shared.registerTargetView(targetId, view: self)
    } else {
      PodverseVideoSurfaceHost.shared.unregisterTargetView(targetId, view: self)
    }
  }

  public override func layoutSubviews() {
    super.layoutSubviews()
    guard let targetId = targetId else { return }
    // Observe size changes (rotation, split view, mini↔full) so the reparented surface fills us.
    PodverseVideoSurfaceHost.shared.layoutSurface(in: self, for: targetId)
  }
}
